Wrong `help` message for `Vec::resize_with` calls that constrain the `Vec` to a wrong type

### Code

```Rust
struct S { xs: Vec<usize> }

fn main() {
    let mut xs = Vec::new();
    xs.resize_with(42, || true);
    S { xs };
}
```


### Current output

```Shell
error[E0308]: mismatched types
 --> <source>:6:9
  |
5 |     xs.resize_with(42, || true);
  |     --                 ------- this argument has type `[closure@<source>:5:24: 5:26]`...
  |     |
  |     ... which causes `xs` to have type `Vec<bool>`
6 |     S { xs };
  |         ^^ expected `Vec<usize>`, found `Vec<bool>`
  |
  = note: expected struct `Vec<usize>`
             found struct `Vec<bool>`
help: use parentheses to call this closure
  |
5 |     xs.resize_with(42, (|| true)());
  |                        +       +++

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
Compiler returned: 1
```


### Desired output

```Shell
error[E0308]: mismatched types
 --> <source>:6:9
  |
5 |     xs.resize_with(42, || true);
  |     --                 ------- this argument has type `[closure@<source>:5:24: 5:26]`...
  |     |
  |     ... which causes `xs` to have type `Vec<bool>`
6 |     S { xs };
  |         ^^ expected `Vec<usize>`, found `Vec<bool>`
  |
  = note: expected struct `Vec<usize>`
             found struct `Vec<bool>`
help: this expression should have type `usize`
  |
5 |     xs.resize_with(42, || true);
  |                           ~~~~

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
Compiler returned: 1
```


### Rationale and extra context

The same `help` message is also shown for function calls and other cases which later lead to type errors, such as returning `xs` from a function with return type `Vec<usize>`.

Present in 1.72.0 stable and `nightly-2023-09-24`:
```console
$ rustc -vV
rustc 1.74.0-nightly (37390d656 2023-09-24)
binary: rustc
commit-hash: 37390d65636dd67e263753a3c04fbc60dcc4348e
commit-date: 2023-09-24
host: x86_64-unknown-linux-gnu
release: 1.74.0-nightly
LLVM version: 17.0.0
```

### Other cases

_No response_

### Anything else?

[Godbolt link](https://godbolt.org/z/EG5dxx8sq)

Reduced to not rely on `Vec`:
```rust
struct S<T>(T);

impl<T> S<T> {
    fn new() -> Self {
        loop {}
    }

    fn constrain<F: Fn() -> T>(&self, _f: F) {}
}

fn main() {
    let s = S::new();
    let c = || true;
    s.constrain(c);
    let _: S<usize> = s;
}
```
```
error[E0308]: mismatched types
  --> src/main.rs:15:23
   |
14 |     s.constrain(c);
   |     -           - this argument has type `[closure@src/main.rs:13:13: 13:15]`...
   |     |
   |     ... which causes `s` to have type `S<bool>`
15 |     let _: S<usize> = s;
   |            --------   ^ expected `S<usize>`, found `S<bool>`
   |            |
   |            expected due to this
   |
   = note: expected struct `S<usize>`
              found struct `S<bool>`
help: use parentheses to call this closure
   |
14 |     s.constrain(c());
   |                  ++
   ```

--- a/compiler/rustc_hir_typeck/src/demand.rs
+++ b/compiler/rustc_hir_typeck/src/demand.rs
@@ -14,7 +14,7 @@ use rustc_middle::ty::adjustment::AllowTwoPhase;
 use rustc_middle::ty::error::{ExpectedFound, TypeError};
 use rustc_middle::ty::fold::BottomUpFolder;
 use rustc_middle::ty::print::with_no_trimmed_paths;
-use rustc_middle::ty::{self, Article, AssocItem, Ty, TypeAndMut, TypeFoldable};
+use rustc_middle::ty::{self, Article, AssocItem, Ty, TypeAndMut, TypeFoldable, TypeVisitableExt};
 use rustc_span::symbol::sym;
 use rustc_span::{BytePos, Span, DUMMY_SP};
 use rustc_trait_selection::infer::InferCtxtExt as _;
@@ -504,12 +504,18 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     // incompatible fix at the original mismatch site.
                     if matches!(source, TypeMismatchSource::Ty(_))
                         && let Some(ideal_method) = ideal_method
+                        && let ideal_arg_ty = self.resolve_vars_if_possible(ideal_method.sig.inputs()[idx + 1])
+                        // HACK(compiler-errors): We don't actually consider the implications
+                        // of our inference guesses in `emit_type_mismatch_suggestions`, so
+                        // only suggest things when we know our type error is precisely due to
+                        // a type mismatch, and not via some projection or something. See #116155.
+                        && !ideal_arg_ty.has_non_region_infer()
                     {
                         self.emit_type_mismatch_suggestions(
                             err,
                             arg_expr,
                             arg_ty,
-                            self.resolve_vars_if_possible(ideal_method.sig.inputs()[idx + 1]),
+                            ideal_arg_ty,
                             None,
                             None,
                         );

--- a/tests/ui/type/type-check/point-at-inference-issue-116155.rs
+++ b/tests/ui/type/type-check/point-at-inference-issue-116155.rs
@@ -1,0 +1,17 @@
+struct S<T>(T);
+
+impl<T> S<T> {
+    fn new() -> Self {
+        loop {}
+    }
+
+    fn constrain<F: Fn() -> T>(&self, _f: F) {}
+}
+
+fn main() {
+    let s = S::new();
+    let c = || true;
+    s.constrain(c);
+    let _: S<usize> = s;
+    //~^ ERROR mismatched types
+}

--- a/tests/ui/type/type-check/point-at-inference-issue-116155.stderr
+++ b/tests/ui/type/type-check/point-at-inference-issue-116155.stderr
@@ -1,0 +1,18 @@
+error[E0308]: mismatched types
+  --> $DIR/point-at-inference-issue-116155.rs:15:23
+   |
+LL |     s.constrain(c);
+   |     -           - this argument has type `{closure@$DIR/point-at-inference-issue-116155.rs:13:13: 13:15}`...
+   |     |
+   |     ... which causes `s` to have type `S<bool>`
+LL |     let _: S<usize> = s;
+   |            --------   ^ expected `S<usize>`, found `S<bool>`
+   |            |
+   |            expected due to this
+   |
+   = note: expected struct `S<usize>`
+              found struct `S<bool>`
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0308`.

